### PR TITLE
[1328] import publish subjects import course

### PR DESCRIPTION
**Pull request by stellenberger**
### Context

- [Trello card](https://trello.com/c/1njDkvXC/1328-m-import-publish-subjects-import-course)
- When courses are retrieved and then imported individually, we want to map a has_many relationship to their respective subjects that are sent with the course, in the format of a subject_code

### Changes proposed in this pull request

- A join table called `CourseSubject` maps the courses and their subjects through references
- Courses `has_many` subjects, subjects `has_many` courses, with `through: course_subjects`
- The courses are only mapped to their subject when they are imported in `TeacherTrainingApi::ImportCourse.call(provider: provider, course: course)`
- Duplicate mappings are skipped

### Guidance to review

- Pre-requisite: Find a legitimate provider through TTAPI and store it in `Provider` locally. Ensure this provider has courses in the current year
- Go into your rails console
```
> provider = Provider.last

> courses = TeacherTrainingApi::RetrieveCourses.call(provider: provider)

> TeacherTrainingApi::ImportCourse.call(provider: provider, course: courses.first)

> CourseSubject.last

> provider.courses.last.subjects
```

### CODE REVIEW

**Comment by stellenberger** [comment]
[[M] Import publish subjects (Import Course)](https://trello.com/c/1njDkvXC/1328-m-import-publish-subjects-import-course)

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 1]
```diff
@@ -9,8 +9,11 @@ module TeacherTrainingApi
       let(:course) { JSON(ApiStubs::TeacherTrainingApi.course) }
       let(:code) { course["attributes"]["code"] }
       let(:name) { course["attributes"]["name"] }
-
+      let(:subject_object) { build(:subject) }
       subject { described_class.call(provider: provider, course: course) }
+      before do
+        allow(Subject).to receive(:where) { [subject_object] }
```
Comment (defong) [inline spec/services/teacher_training_api/import_course_spec.rb:15]:
```suggestion
        allow(Subject).to receive(:where).with({code:[subject_object.code] }).and_return([subject_object])
```

Don't be loose...

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 2]
```diff
@@ -22,6 +25,16 @@ module TeacherTrainingApi
...
+
+          it "creates a join table between the course and subject" do
+            expect { subject }.to change { CourseSubject.count }.by(1)
+            expect(provider.courses.find_by(code: code, name: name).subjects.length).to eq 1
+          end
+
+          it "has accessors to course subjects" do
+            subject
+            expect(provider.courses.find_by(code: code, name: name).subjects.last).to eq subject_object
+          end
```
Comment (defong) [inline spec/services/teacher_training_api/import_course_spec.rb:37]:
```suggestion
          it "creates the course with the correct subjects" do
            expect { subject }.to change { CourseSubject.count }.from(0).to(1)
            expect(provider.courses.find_by(code: code, name: name).subjects).to match [subject_object]
          end
```

`creates a join table between the course and subject` reads odd when considering that's its a model words

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 3]
```diff
@@ -40,6 +57,11 @@ module TeacherTrainingApi
...
             expect { subject }.not_to(change { Course.count })
           end
+
+          it "does not create a duplicate course_subject join" do
+            subject
+            expect { subject }.not_to(change { CourseSubject.count })
+          end
         end
 
         context "with a different name" do
```
Comment (defong) [inline spec/services/teacher_training_api/import_course_spec.rb:67]: If name can change needs a test to show:

- subjects will not be duplicated
- subjects can be updated

File: spec/factories/course_subjects.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
Comment (defong) [inline spec/factories/course_subjects.rb:3]: No needed

You would want to go to courses.rb and do

```ruby
    factory :course_with_a_subject do
      subjects { [association(:subject)] }
    end
```


https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#has_many-associations

**Review by defong — CHANGES_REQUESTED** [review]
Why are you taking a :hankey:?

:champagne: `dump.rdb`

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 5]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
-          it "does not creates a duplicate course" do
-            expect { subject }.not_to(change { Course.count })
+                 name: course_name,
+                 subjects: course_subjects)
+        end
+
+        let(:course_name) { imported_name }
+        before { existing_course }
+
+        context "imported course attributes was not changed" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:69]: It's quite nice when these `context` string, 'string' together to form a sentence.

Before, the test output read like this:

```
"when the course code already exists for that provider" -> "with the same name" -> it "does not creates a duplicate course"
```

Now it reads like:

```
"when the course code already exists for that provider" -> "imported course attributes was not changed" -> "with the same name" -> it "does not creates a duplicate course"
```

So it's lost its readability.

What are we trying to achieve with these new contexts? They don't wrap any new `lets`. As such, I'd be inclined to remove `context "imported course attributes was not changed"` and `context "imported course attributes was changed"`

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 6]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
+          context "with the same name" do
+            it "does not creates a duplicate course" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:71]:
```suggestion
            it "does not create a duplicate course" do
```

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 7]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
+              expect { subject }.not_to(change { Course.count })
+            end
+          end
+
+          context "with the same subjects" do
+            it "does not creates a duplicate course subjects" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:77]:
```suggestion
            it "does not create duplicate course subjects" do
```

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 8]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
 
-          it "updates the name" do
-            subject
-            expect(provider.courses.find_by(code: code).name).to eq name
+            it "updates the name" do
+              subject
+              expect(provider.courses.find_by(code: imported_code).name).to eq(imported_name)
+            end
+
+            it "does not creates a duplicate course" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:92]:
```suggestion
            it "does not create a duplicate course" do
```

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 9]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
+              expect { subject }.not_to(change { Course.count })
+            end
+
+            it "does not creates a duplicate course subjects" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:96]:
```suggestion
            it "does not create duplicate course subjects" do
```

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 10]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
+            let(:course_subjects) { [music_subject] }
+
+            let(:music_subject) { create(:subject, :music) }
+
+            it "updates the subjects" do
+              subject
+              expect(provider.courses.find_by(code: imported_code).subjects).to eq course_subjects
+            end
+
+            it "does not creates a duplicate course" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:111]:
```suggestion
            it "does not create a duplicate course" do
```

File: spec/services/teacher_training_api/import_course_spec.rb [hunk 11]
```diff
@@ -20,40 +31,96 @@ module TeacherTrainingApi
...
+              expect { subject }.not_to(change { Course.count })
+            end
+
+            it "does not creates a duplicate course subjects" do
```
Comment (AbigailMcP) [inline spec/services/teacher_training_api/import_course_spec.rb:115]:
```suggestion
            it "does not create duplicate course subjects" do
```

**Review by AbigailMcP — COMMENTED** [review]
Looks good to me! Just a couple of comments tidying up the spec file 👌

File: spec/services/teacher_training_api/import_course_spec.rb (hunk 5 above)
Comment (defong) [reply spec/services/teacher_training_api/import_course_spec.rb]: Fixed, Monday was a Non-English day :sweat_smile:

**Review by AbigailMcP — APPROVED** [review]

**Pull request merged**